composer.json

```diff
@@ -58,7 +58,9 @@
     },
     "suggest": {
         "ext-dom": "For handling output formats in XML",
-        "ext-mbstring": "For handling non-UTF8 characters."
+        "ext-mbstring": "For handling non-UTF8 characters.",
+        "doctrine/annotations": "Required (^2) if you want to use @DoctrineAnnotation rule set (or single fixers related to Doctrine Annotations)",
+        "doctrine/lexer": "Required (^2 || ^3) if you want to use @DoctrineAnnotation rule set (or single fixers related to Doctrine Annotations)"
```

Comment:
I'm very mixed here.

Imagine the following:

- as Dev, I use 3.23.0 Fixer, and having `@DoctrineAnnotation` enabled for my nproject
- as Fixer maintainer, we drop `doctrine/*` dependency and release 3.24.0
- as Dev, I upgrade to 3.24.0 and... boom - I have error running the tool

effect? anyone with ^3 who is using `@DoctrineAnnotation` rules will start have their jobs (eg CI) failing out of the blue.

yeah, we could hope that they have those deps already present - but if not, eg because they install Fixer in dedicated folder (eg dev-tools/), then we are messing up with their current CI pipelines and local execution.

Reply:
Yes, I understand. But:
- we could include explicit information about it in the release notes
- fix on a Dev side is easy, because there would be and descriptive error with information that these deps have to be installed

I don't see other resolution for #7250, we just need to make these deps optional. Community goes more and more towards attributes and installing those 2 deps because some unused fixers require them is not OK.

Reply:
Also, I believe it's pretty much like dropping PHP version support in minor version - API is unchanged, but it may require some action on the users' side, to continue with newer version.

Reply:
I don't agree it;s like dropping PH version support in minor version.

Imagine:
- I use PHP v10.0 and PHP CS Fixer v20.0
- PHP CS Fixer v20.1 start requiring PHP v11+
- my CI is still installing v20.0, it's not bumping Fixer to v20.1

on dropping the required packages to have all rules working, the CI will do the Fixer upgrade and then fail on runtime [for those who use related rules]

it's not about how easy or not it is to apply needed changes on userland, it's how unexpected crash on runtime they will face, with unattended tool upgrade ('hey, it's minor, we allow all ^3')


I see the value of proposed change, yet rather on the side to call it a BC breaker, due to mentioned reasons.

Reply:
for best case, and let's say for v4, we want to have the opt-in. So user can decide to install additional doctrine libs.

is it anyhow possible to have opt-out? so by default we do install those doctrine libs, but somehow on composer level, user can declare to skip them?

Reply:
> is it anyhow possible to have opt-out? so by default we do install those doctrine libs, but somehow on composer level, user can declare to skip them?

I am not aware of such approach. If we keep strict `require` then these packages will be installed regardless if they are used or not. This has 2 drawbacks already mentioned: installing superfluous packages and possible package resolution conflict (like in our case, we're stuck on 3.14).

> it's not about how easy or not it is to apply needed changes on userland, it's how unexpected crash on runtime they will face, with unattended tool upgrade ('hey, it's minor, we allow all ^3')

I think I disagree. In general, in the userland, when you install such tool you rather pin it to exact version and just don't install latest version in the CI. So each run in the CI works until you run `composer update` or bots like Renovate does it for you. In any way, it should go through merge request and if project uses `@DoctrineAnnotation` ruleset or any of the rules from this set, this would be moment when you may encounter the problem. I believe if we include proper information in the release notes, then people can easily fix it when upgrading, because currently Doctrine packages are already installed as a strict dependency, so the only thing you need to do is add requirement on the project's side to continue to use them. And this is why I said it's like dropping PHP version - you need to align requirements to use new version, or stay with the one that currently works properly 🙂.

Reply:
@keradus your scenario implies that a project has enabled the DoctrineAnnotation ruleset without anything depending on doctrine/annotations in the project though, which would mean that they don't actually use them

Reply:
You can have the ruleset enabled without it doing anything. Previous version of CS fixer would no-op, new version would crash, IIUC?

Reply:
If you're using a PHAR, you might not have any place to "require" this from.

Reply:
If the concern is with PHP-CS-Fixer dropping a required dependency that a downstream consumer requires in their application, and that dependency is only installed as a transient dependency (AKA another package requires it), and updating this tool breaks their application by uninstalling the transient dependency their application needs, then it's a fair concern to be aware of but IMO it's not the big B/C break you're arguing it to be.

The `@DoctrineAnnotation` ruleset is already an optional thing, it's not enabled by default with a fresh install of this package or through tools such as Laravel Pint or the Symfony Flex recipe.

This tool SHOULD be a `require-dev` dependency for Composer-managed installs.  So for applications using annotations for their production code and this tool in that configuration, either the application is already correctly requiring `doctrine/annotations` (which has always required `doctrine/lexer`, so it's not possible to somehow end up in a state where the Annotations package is installed but not the Lexer package), or the app is lucking out and something else in its production dependency chain is installing the Annotations package.  For folks installing the PHAR, as noted in this PR, these optional Doctrine dependencies would still be packaged in that distributable, so they wouldn't be affected by this.  Which means the users with legitimate impact would be:

- Users with this tool as a `require` dependency in their root application and relying on it to provide the transient dependencies to either the Annotations or Lexer packages
- Users with this tool installed in a separate path from their root application (i.e. `tools/php-cs-fixer/composer.json`) with the `@DoctrineAnnotation` ruleset enabled but not requiring `doctrine/annotations` in the tool manifest

To me, this looks like a very low-risk change, with a one-line mitigation for the folks actually affected by it: `composer require doctrine/annotations`.  I can understand wanting to be very cautious with B/C, but I personally wouldn't consider "applications depending on a transient dependency provided by us" as a case that needs to be covered under a B/C promise.

Reply:
This seems like a BC break for sure, we can make it a smaller break if we skip the rules instead of crashing, but that introduces own issues with UX (suddenly you have no-op rules, how does the user know that?)

Reply:
Installation permuatations
1. CS fixer is a dev dep of the project
2. CS fixer is installed in a separate folder as a utility
3. CS fixer is used via PHAR (directly or via Docker, say via `jakzal/phpqa`)

Deps
1. you have annotations installed in your own app and use them
2. you have annotations installed in your own app, but don't use them

Enabled rules:
1. you have Doctrine ruleset enabled and it's doing something
2. you have Doctrine ruleset enabled and it's doing nothing
3. you don't have Doctrine ruleset enabled

There's a lot of permutations here and IMO there should be a consideration what happens in each one.

Reply:
> This seems like a BC break for sure, we can make it a smaller break if we skip the rules instead of crashing, but that introduces own issues with UX (suddenly you have no-op rules, how does the user know that?)

We could at least add a warning to the output.